tests/test_nn.py

```diff
@@ -108,3 +108,120 @@ def test_mlp(getkey):
+    x = jrandom.normal(getkey(), (1, 32))
+    assert conv(x).shape == (3, 30)
+
+    # Some keyword arguments
+    conv = eqx.nn.Conv1d(1, out_channels=3, kernel_size=(3,), key=getkey())
+    x = jrandom.normal(getkey(), (1, 32))
+    assert conv(x).shape == (3, 30)
+
+    # All keyword arguments
+    conv = eqx.nn.Conv1d(
+        in_channels=1,
+        out_channels=3,
+        kernel_size=(3,),
+        padding=1,
+        use_bias=False,
+        key=getkey(),
+    )
+    x = jrandom.normal(getkey(), (1, 32))
+    assert conv(x).shape == (3, 32)
+
+    # Test strides
+    conv = eqx.nn.Conv1d(
+        in_channels=3,
+        out_channels=1,
+        kernel_size=(3,),
+        stride=2,
+        padding=1,
+        use_bias=True,
+        key=getkey(),
+    )
+    x = jrandom.normal(getkey(), (3, 32))
+    assert conv(x).shape == (1, 16)
+
+
+def test_conv2d(getkey):
+    # Positional arguments
+    conv = eqx.nn.Conv2d(1, 3, 3, key=getkey())
+    x = jrandom.normal(getkey(), (1, 32, 32))
+    assert conv(x).shape == (3, 30, 30)
+
+    # Some keyword arguments
+    conv = eqx.nn.Conv2d(1, out_channels=3, kernel_size=(3, 3), key=getkey())
+    x = jrandom.normal(getkey(), (1, 32, 32))
+    assert conv(x).shape == (3, 30, 30)
+
+    # All keyword arguments
+    conv = eqx.nn.Conv2d(
+        in_channels=1,
+        out_channels=3,
+        kernel_size=(3, 3),
+        padding=1,
+        use_bias=False,
+        key=getkey(),
+    )
+    x = jrandom.normal(getkey(), (1, 32, 32))
+    assert conv(x).shape == (3, 32, 32)
+
+    # Test strides
+    conv = eqx.nn.Conv2d(
+        in_channels=3,
+        out_channels=1,
+        kernel_size=(3, 3),
+        stride=2,
+        padding=1,
+        use_bias=True,
+        key=getkey(),
+    )
+    x = jrandom.normal(getkey(), (3, 32, 32))
+    assert conv(x).shape == (1, 16, 16)
+
+
+def test_conv3d(getkey):
+    # Positional arguments
+    conv = eqx.nn.Conv3d(1, 3, 3, key=getkey())
+    x = jrandom.normal(getkey(), (1, 3, 32, 32))
+    assert conv(x).shape == (3, 1, 30, 30)
+
+    # Some keyword arguments
+    conv = eqx.nn.Conv3d(1, out_channels=3, kernel_size=(3, 3, 3), key=getkey())
+    x = jrandom.normal(getkey(), (1, 3, 32, 32))
+    assert conv(x).shape == (3, 1, 30, 30)
+
+    # All keyword arguments
+    conv = eqx.nn.Conv3d(
+        in_channels=1,
+        out_channels=3,
+        kernel_size=(3, 3, 3),
+        padding=1,
+        use_bias=False,
+        key=getkey(),
+    )
+    x = jrandom.normal(getkey(), (1, 3, 32, 32))
+    assert conv(x).shape == (3, 3, 32, 32)
+
+    # Test strides
+    conv = eqx.nn.Conv3d(
+        in_channels=3,
+        out_channels=1,
+        kernel_size=(3, 3, 3),
+        stride=2,
+        padding=1,
+        use_bias=True,
+        key=getkey(),
+    )
+    x = jrandom.normal(getkey(), (3, 3, 32, 32))
+    assert conv(x).shape == (1, 2, 16, 16)
+
+
+def test_conv(getkey):
+    test_conv1d(getkey)
+    test_conv2d(getkey)
+    test_conv2d(getkey)
```

Comment:
This will actually mean that each `test_conv*d` will be run twice. pytest automatically detects all functions whose names begin with `test`, so it'll pick up on `test_conv`, `test_conv1d`, `test_conv2d` and `test_conv3d` separately.

In this case I think all that needs to happen is to delete `test_conv`.

---

Besides that: I think it'd be good to explicitly test the output value (and not just the shape) in just a couple of very simple cases, just as a sanity check. Something like:

```python
conv = Conv1d(...)
new_weight = jnp.array(...)
data = jnp.array(...)
assert new_weight.shape == conv.weight.shape  # make sure the test itself is valid
conv = eqx.tree_at(lambda x: x.weight, conv, new_weight)  # manually set the weight
assert jnp.allclose(conv(data), ...)
```

Reply:
Latest commit now checks output values and removed the unnecessary test_conv function - let me know if the tests look good!